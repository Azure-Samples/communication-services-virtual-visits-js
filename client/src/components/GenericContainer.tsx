// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import MobileDetect from 'mobile-detect';
import { LayerHost, Stack, Theme } from '@fluentui/react';
import { getGenericContainerStyles } from '../styles/GenericContainer.styles';
import { getDefaultLayerHostStyles } from '../styles/Common.styles';
import React from 'react';

interface GenericContainerProps {
  layerHostId: string;
  theme: Theme;
}

const GenericContainer = (props: React.PropsWithChildren<GenericContainerProps>): JSX.Element => {
  const { layerHostId, theme } = props;
  const containerStyles = getGenericContainerStyles(theme, new MobileDetect(window.navigator.userAgent).mobile());

  return (
    <LayerHost id={layerHostId} style={getDefaultLayerHostStyles()}>
      <Stack styles={containerStyles} tokens={{ childrenGap: 15 }}>
        {props.children}
      </Stack>
    </LayerHost>
  );
};

export default GenericContainer;
