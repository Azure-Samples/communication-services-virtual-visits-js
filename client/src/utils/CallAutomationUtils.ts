// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { CommunicationUserIdentifier } from '@azure/communication-common';
import { CallAdapterState, CommonCallAdapter } from '@azure/communication-react';

export type SummarizeResult = {
  recap: string;
  chapters: {
    chapterTitle: string;
    narrative: string;
  }[];
};

export type CallTranscription = {
  text: string;
  confidence: number;
  participant: CommunicationUserIdentifier;
  resultState: 'intermediate' | 'final';
}[];

/**
 * Fetches the transcription for the given server call ID.
 * @param serverCallId - The server call ID to fetch the transcription for.
 * @returns - the CallTranscription object containing the transcription data.
 */
export const fetchTranscript = async (serverCallId: string): Promise<CallTranscription> => {
  const response = await fetch(`/api/fetchTranscript`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    throw new Error('Failed to fetch transcript: ' + response.statusText);
  }

  return ((await response.json()) as { transcript: CallTranscription }).transcript;
};

/**
 * fetches the transcription status for the given server call ID. This is called when
 * the user first joins the call to check if transcription is already started.
 * @param serverCallId - The server call ID to fetch the transcription status for.
 */
export const fetchTranscriptionStatus = async (serverCallId: string): Promise<void> => {
  const response = await fetch(`/api/fetchTranscriptionState`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    throw new Error('Failed to fetch transcription status: ' + response.statusText);
    return;
  }
  return;
};

/**
 * Starts the transcription for the call.
 * @param serverCallId - The server call ID to start transcription for.
 * @param transcriptionOptions - The options for transcription, including locale.
 * @returns - True if the transcription was started successfully, false otherwise.
 */
export const startTranscription = async (
  serverCallId: string,
  transcriptionOptions?: {
    locale?: string;
  }
): Promise<boolean> => {
  console.log('Starting transcription for call:', serverCallId);
  const response = await fetch('/api/startTranscription', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId,
      options: transcriptionOptions
    })
  });
  if (!response.ok) {
    console.error('Failed to start transcription:', response);
    return false;
  }
  console.log('Started transcription:', transcriptionOptions);
  return true;
};

/**
 * Stops the transcription for the call.
 * @param serverCallId - The server call ID to stop transcription for.
 * @returns {boolean} - True if the transcription was stopped successfully, false otherwise.
 */
export const stopTranscription = async (serverCallId: string): Promise<boolean> => {
  console.log('Stopping transcription for call:', serverCallId);

  const response = await fetch('/api/stopTranscription', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId
    })
  });
  if (!response.ok) {
    console.error('Failed to stop transcription:', response);
    return false;
  }
  console.log('Stopped transcription');
  return true;
};

/**
 * Connect the Call Automation service to the call.
 * @param callAdaterState - The call adapter state to check for the call.
 * @returns
 */
export const connectToCallAutomation = async (callAdaterState: CallAdapterState): Promise<boolean> => {
  if (callAdaterState.call?.info !== undefined && callAdaterState.call?.state === 'Connected') {
    const serverCallID = await callAdaterState.call.info.getServerCallId();
    const response = await fetch('/api/connectRoomsCall', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        serverCallId: serverCallID
      })
    });
    if (!response.ok) {
      throw new Error('Failed to start call with transcription');
    }
    return true;
  }
  return false;
};

/**
 * Fetch the call summary from the server.
 * @param adapter - The call adapter to get the call ID from.
 * @param locale - locale code for the language to be used in the summary.
 * @returns summary result if there is a summary available.
 */
export const getCallSummaryFromServer = async (
  adapter: CommonCallAdapter,
  locale: LocaleCode
): Promise<SummarizeResult> => {
  console.log('Getting summary from server...');
  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const serverCallId = await adapter.getState().call?.info?.getServerCallId();
    if (!serverCallId) {
      console.error('Call ID not found');
      throw new Error('Call ID not found');
    }

    const response = await fetch('/api/summarizeTranscript', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({ serverCallId: serverCallId, locale: locale })
    });

    if (!response.ok) {
      console.error('Response Failed: ', response.statusText);
      throw new Error('Summarization request failed');
    }

    const result = await response.json();
    console.log('Summary result:', result);
    return result as SummarizeResult;
  } catch (error) {
    console.error('Error fetching summary:', error);
    throw error;
  }
};

/**
 * Update the participants in the call record for transcription.
 * @param participant - The participant to add to the call record.
 */
export const sendParticipantInfoToServer = async (
  participant: { userId: string; displayName: string },
  serverCallId?: string
): Promise<void> => {
  if (!serverCallId) {
    console.error('Server call ID not found');
    return;
  }
  const response = await fetch('/api/updateParticipants', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId: serverCallId,
      participant: participant
    })
  });
  if (!response.ok) {
    console.error('Failed to update participants:', response);
  }
};

/**
 * Fetches the participants in the call. This is needed for when we are computing the transcript for download.
 * @param serverCallId - The server call ID to fetch the participants for.
 * @returns display name and user ID of the participants in the call.
 */
export const fetchParticipants = async (serverCallId: string): Promise<{ userId: string; displayName: string }[]> => {
  const response = await fetch('/api/fetchParticipants', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      serverCallId: serverCallId
    })
  });
  if (!response.ok) {
    console.error('Failed to fetch participants:', response);
    return [];
  }
  return ((await response.json()) as { participants: { userId: string; displayName: string }[] }).participants;
};

/**
 * Language codes used in the localization of the components.
 */
export type LocaleCode =
  | 'ar-AE'
  | 'ar-SA'
  | 'da-DK'
  | 'de-DE'
  | 'en-AU'
  | 'en-CA'
  | 'en-GB'
  | 'en-IN'
  | 'en-NZ'
  | 'en-US'
  | 'es-ES'
  | 'es-MX'
  | 'fi-FI'
  | 'fr-CA'
  | 'fr-FR'
  | 'hi-IN'
  | 'it-IT'
  | 'ja-JP'
  | 'ko-KR'
  | 'nb-NO'
  | 'nl-BE'
  | 'nl-NL'
  | 'pl-PL'
  | 'pt-BR'
  | 'ru-RU'
  | 'sv-SE'
  | 'zh-CN'
  | 'zh-HK'
  | 'cs-CZ'
  | 'pt-PT'
  | 'tr-TR'
  | 'vi-VN'
  | 'th-TH'
  | 'he-IL'
  | 'cy-GB'
  | 'uk-UA'
  | 'el-GR'
  | 'hu-HU'
  | 'ro-RO'
  | 'sk-SK'
  | 'zh-TW';

/**
 * Mapping of locale codes to their display names.
 */
export const localeDisplayNames: Record<LocaleCode, string> = {
  'ar-AE': 'Arabic - U.A.E.',
  'ar-SA': 'Arabic - Saudi Arabia',
  'da-DK': 'Danish',
  'de-DE': 'German - Germany',
  'en-AU': 'English - Australia',
  'en-CA': 'English - Canada',
  'en-GB': 'English - United Kingdom',
  'en-IN': 'English - India',
  'en-NZ': 'English - New Zealand',
  'en-US': 'English - United States',
  'es-ES': 'Spanish - Spain (Modern Sort)',
  'es-MX': 'Spanish - Mexico',
  'fi-FI': 'Finnish',
  'fr-CA': 'French - Canada',
  'fr-FR': 'French - France',
  'hi-IN': 'Hindi',
  'it-IT': 'Italian - Italy',
  'ja-JP': 'Japanese',
  'ko-KR': 'Korean',
  'nb-NO': 'Norwegian (Bokm√•l)',
  'nl-BE': 'Dutch - Belgium',
  'nl-NL': 'Dutch - Netherlands',
  'pl-PL': 'Polish',
  'pt-BR': 'Portuguese - Brazil',
  'ru-RU': 'Russian',
  'sv-SE': 'Swedish',
  'zh-CN': "Chinese - People's Republic of China",
  'zh-HK': 'Chinese - Hong Kong SAR',
  'cs-CZ': 'Czech',
  'pt-PT': 'Portuguese - Portugal',
  'tr-TR': 'Turkish',
  'vi-VN': 'Vietnamese',
  'th-TH': 'Thai',
  'he-IL': 'Hebrew',
  'cy-GB': 'Welsh',
  'uk-UA': 'Ukrainian',
  'el-GR': 'Greek',
  'hu-HU': 'Hungarian',
  'ro-RO': 'Romanian',
  'sk-SK': 'Slovak',
  'zh-TW': 'Chinese - Taiwan'
};
