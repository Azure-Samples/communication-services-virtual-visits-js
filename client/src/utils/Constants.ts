// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

export const BOOKINGS_README_URL = 'https://aka.ms/virtual-appointments-sample-bookings';
export const BOOKINGS_SPECIMEN_URL =
  'https://microsoftbookings.azurewebsites.net/?organization=financialservices&UICulture=en-US';
