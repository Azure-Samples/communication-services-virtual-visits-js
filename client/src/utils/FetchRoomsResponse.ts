// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import createError from 'http-errors';
import { CreateRoomResponse, JoinRoomResponse } from '../models/RoomModel';

export const fetchRoomsResponse = async (roomId: string, userId: string): Promise<JoinRoomResponse | undefined> => {
  const response = await fetch('/api/rooms/token', {
    method: 'POST',
    body: JSON.stringify({
      roomId: roomId,
      userId: userId
    }),
    headers: {
      'content-type': 'application/json;charset=UTF-8'
    }
  });

  if (response.status !== 200) {
    throw new createError(
      response.status,
      `Unable to retrieve user token. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }

  const responseContent = await response.text();
  const roomResponse = JSON.parse(responseContent);
  return roomResponse;
};

export const createRoom = async (): Promise<CreateRoomResponse> => {
  const response = await fetch('/api/rooms', {
    method: 'POST',
    headers: {
      'content-type': 'application/json;charset=UTF-8'
    }
  });
  if (response.status !== 201) {
    throw new createError(
      response.status,
      `Unable to create room. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }
  const responseContent = await response.text();
  const roomsResponse = JSON.parse(responseContent);
  return roomsResponse;
};
