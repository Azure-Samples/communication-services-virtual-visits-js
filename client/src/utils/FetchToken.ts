// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { CommunicationUserToken } from '@azure/communication-identity';
import createError from 'http-errors';

export async function fetchToken(): Promise<CommunicationUserToken | undefined> {
  const response = await fetch('/api/token', { method: 'GET' });

  if (response.status !== 200) {
    throw new createError(
      response.status,
      `Unable to retrieve config. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }

  const responseContent = await response.text();
  const token = JSON.parse(responseContent, parseDate);

  return token;
}

function parseDate(key: string, value: string): Date | string {
  if (key === 'expiresOn') {
    return new Date(value);
  }
  return value;
}
