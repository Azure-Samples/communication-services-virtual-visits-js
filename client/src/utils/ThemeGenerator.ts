// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { ThemeGenerator, themeRulesStandardCreator, createTheme, BaseSlots } from '@fluentui/react';
import { Theme } from '@fluentui/theme';

export function generateTheme(primaryColor: string): Theme {
  const themeRules = themeRulesStandardCreator();
  ThemeGenerator.setSlot(themeRules[BaseSlots[BaseSlots.primaryColor]], primaryColor, false, true, true);
  ThemeGenerator.insureSlots(themeRules, false);
  const themeAsJson = ThemeGenerator.getThemeAsJson(themeRules);
  return createTheme({ ...{ palette: themeAsJson } });
}
