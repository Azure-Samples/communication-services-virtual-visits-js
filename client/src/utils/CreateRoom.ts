// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { RoomParticipant, RoomParticipantRole } from '../models/RoomModel';
import { createRoom } from './FetchRoomsResponse';
import { makeRoomsJoinUrl } from './GetMeetingLink';

export const createRoomAndRedirectUrl = async (): Promise<string> => {
  let redirectUrl: string;
  try {
    const roomResponse = await createRoom();
    const roomId = roomResponse.roomId;
    const userId = roomResponse.participants.find(
      (participant: RoomParticipant) => (participant.role as RoomParticipantRole) === RoomParticipantRole.presenter
    )?.id;
    if (userId) redirectUrl = makeRoomsJoinUrl(roomId, userId);
    else throw new Error(`room does not have participant with role ${RoomParticipantRole.presenter}`);
  } catch (e) {
    //EndUser to add logging errors in place of console logging below
    console.log(e);
    throw e;
  }
  return redirectUrl;
};
