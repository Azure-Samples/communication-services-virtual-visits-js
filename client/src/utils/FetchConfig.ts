// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import { AppConfigModel } from '../models/ConfigModel';
import { generateTheme } from './ThemeGenerator';
import createError from 'http-errors';

export async function fetchConfig(): Promise<AppConfigModel | undefined> {
  const response = await fetch('/api/config', { method: 'GET' });
  if (response.status !== 200) {
    throw new createError(
      response.status,
      `Unable to retrieve config. Status: ${response.status}. Message: ${response.statusText}.`
    );
  }

  const responseContent = await response.text();
  const config = JSON.parse(responseContent);
  if (!config) {
    throw new Error(`Received empty config response.`);
  }

  const appConfigModel: AppConfigModel = {
    ...config,
    theme: generateTheme(config.colorPalette)
  };

  return appConfigModel;
}
