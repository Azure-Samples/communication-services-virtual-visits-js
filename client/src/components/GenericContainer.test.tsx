// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

import * as renderer from 'react-test-renderer';
import { generateTheme } from '../utils/ThemeGenerator';
import GenericContainer from './GenericContainer';

describe('DefaultContainer', () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it.each([['desktop'], ['mobile']])('uses correct styles if form factor is %s', async (formFactor: string) => {
    const theme = generateTheme('#FFFFFF');

    const userAgent =
      formFactor === 'mobile'
        ? 'Mozilla/5.0 (iPhone; CPU iPhone OS 14_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Mobile/15E148 Safari/604.1'
        : 'Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:47.0) Gecko/20100101 Firefox/47.0';

    const userAgentGetter = jest.spyOn(window.navigator, 'userAgent', 'get');
    userAgentGetter.mockReturnValue(userAgent);

    const container = renderer
      .create(
        <GenericContainer layerHostId="layerHost" theme={theme}>
          <div>Hello, World!</div>
        </GenericContainer>
      )
      .toJSON();

    expect(container).toMatchSnapshot();
  });
});
