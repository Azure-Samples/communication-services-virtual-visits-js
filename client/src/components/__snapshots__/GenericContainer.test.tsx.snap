// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DefaultContainer uses correct styles if form factor is desktop 1`] = `
<div
  className="ms-LayerHost"
  id="layerHost"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "position": "relative",
    }
  }
>
  <div
    className="ms-Stack css-109"
  >
    <div>
      Hello, World!
    </div>
  </div>
</div>
`;

exports[`DefaultContainer uses correct styles if form factor is mobile 1`] = `
<div
  className="ms-LayerHost"
  id="layerHost"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "position": "relative",
    }
  }
>
  <div
    className="ms-Stack css-110"
  >
    <div>
      Hello, World!
    </div>
  </div>
</div>
`;
